File: client/broker/Broker.py
#!/usr/bin/env python
from random import *
from socket import *
from Message import *
from SocketWorker import *
from BrokerCallBack import *
from SessionManager import *

"""
The Broker manages the communcation between different objects (software components) that are locally separated or even on
the same machine. It uses a registrar to authenticate before the actual communication takes place, the registrars IP
needs to be known, the all other communication partner are dynamically assigned.
"""
class Broker:

    def __init__(self):
        self._objectID = 0x80000000
        self._sessionID = 0x80000000
        self._requestID = 0x80000000
        self._clientPort = 0
        self._clientAddress = ""
        self._registrarAddress = "127.0.0.1"
        self._registrarPort = 0x4931
        self._serverAddress = ""
        self._serverPort = 0
        self._serverName = ''
        self._socket = socket(AF_INET, SOCK_DGRAM)
        self._callBack = 0
        self._username = ""
        self._password = ""
        self._sessionManager = 0
        self._state = "offline"
        self._externalIP = ''
        self._externalPort = 0

    """
    This method needs to be called to initialize the communication with the registrar
    and to enable the Broker to communicate with other objects.
    """
    def init(self):
        self._clientPort = randint(1025,65350)
        self._socket.bind((self._clientAddress,self._clientPort))
        sw = SocketWorker(self)
        sw.start()

        # load SessionManager
        self._sessionManager = SessionManager(self)
        self._sessionManager.start()
        
        print "[Broker] SocketWorker started and back in main thread."

    """
    The method assigns the username and password needed to authenticate at the registrar
    """
    def setAuthenticationData(self, username, password):
        self._username = username
        self._password = password


    """
    The method stores a callback object at the Broker, it will get informed about
    received Messages.
    bcb BrokerCallBack object
    """
    def registerCallBack(self, bcb):
        self._callBack = bcb
        bcb._broker = self

    """
    The method returns the Broker state, it can be either "offline" or "online", online indicates that an active session
    is existing.
    """
    def getState(self):
        return self._state

    def setServerName(self, servername):
        self._serverName = servername

    """
    The method sends a message to the server. in case of an incomplete header it will complete it.
    """
    def send(self,message,destination='server'):

        message._sessionID = self._sessionID

        if (message._objectID == 0):
            message._objectID = self._objectID

        if message._requestID == 0:
            self._requestID += 1
            self._requestID = self._requestID&0x7FFFFFFF
            message._requestID = self._requestID

        if destination == 'server' and self._serverAddress != "":
            self._socket.sendto(message.encodeBER(), (self._serverAddress,self._serverPort))
        else:
            self._socket.sendto(message.encodeBER(), (self._registrarAddress,self._registrarPort))

File: server/broker/Broker.java
package broker;

import broker.object.BrokerCallBack;
import broker.service.com.*;
import broker.service.com.protocol.*;
import broker.service.index.*;

/**
 * Singleton wrapper facade for the different BrokerServices.
 */
public class Broker {

	/**
	 * static singleton instance
	 */
	private static Broker instance = null;
	
	/**
	 * The NetworkService used by the Broker
	 */
	private NetworkService netService = null;
	
	/**
	 * The ProtocolService used by the Broker
	 */
	private ProtocolService proService = null;

	/**
	 * The CommunicationService used by the Broker
	 */
	private CommunicationService comService = null;

	/**
	 * The CommunicationService used by the Broker
	 */
	private MessageRouterService routService = null;
	
	/**
	 * The IndexService used by the Broker
	 */
	private IndexService indexService = null;
	

	
	/** 
	 * Private constructor, to provide singleton object
	 */
	private Broker() {

		/* IndexService */
		this.indexService = IndexService.getInstance();

		/* Message Router Service */
		this.routService = new MessageRouterService(this.indexService);

		/* Communication Service */
		this.comService = new CommunicationService(this.indexService);
		this.comService.setNextService(this.routService);

		/* Protocol Service */
		this.proService = new ProtocolService();
		this.proService.setNextService(this.comService);

		/* NetworkService */
		this.netService = new NetworkService(this.indexService);
		this.netService.setNextService(this.proService);
		
	}
	
	/**
	 * This method is a synchronized way to create or access a singleton instance of Broker.
	 * @return singleton instance of Broker
	 */
	public synchronized static Broker getInstance() {
	
		if ( Broker.instance == null )
			Broker.instance = new Broker();
	
		return Broker.instance;
	
	}
	
	/**
	 * This method needs to be called to initialize the communication with the registrar
	 * and to enable the Broker to communicate with other objects.
	 */
	public void init() {
	
		/* send ObjectHello to registrar, the rest of the communication will be handled in the CommunicationService */
		Message m = new Message(0x80000000,0x80000000,0x80000000);
		this.send(m);
		
	}
	
	/**
	 * The method stores a callback object at the Broker, it will get informed about
	 * received Messages and in case of joining or leaving objects.
	 * @param bcb BrokerCallBack object
	 */
	public void registerCallBack(BrokerCallBack bcb) {
	
		bcb.setBroker(this);
		this.indexService.registerBrokerCallBack(bcb);
	
	}
	
	/**
	 * The method sends the Message to it's receiver, the destination will be determined by the
	 * objectID field in the Message header.
	 * @param m, Message that shall be send
	 */
	public void send(Message m) {
	
		if ( this.routService != null )
			this.routService.sapUpperLayer(m);
	
	}
	
	/**
	 * The method sends the Message to all available receiver objects, this is used to inform
	 * all objects about changes in one object, therefore the objectID shall contain the value
	 * of the object that changed.
	 * @param m, Message that shall be broadcasted
	 */	
	public void sendBroadcast(Message m) {
	
		if ( this.routService != null ) {
		
			m.setRequestID(0x7FFFFFFF);
			this.routService.sapUpperLayer(m);
			
		}
	
	}
	
	/**
	 * The method sets the authentication data needed to authenticate against the registrar.
	 * @param username name of the user
	 * @param password user password
	 */
	public void setAuthenticationData(String username, String password) {
	
		this.indexService.setAuthenticationData(username,password);
	
	}
	
}
